apps/remote_control/lib/lexical/remote_control/code_intelligence/rename.ex

```diff
@@ -0,0 +1,178 @@
+defmodule Lexical.RemoteControl.CodeIntelligence.Rename do
+  alias Lexical.Ast
+  alias Lexical.Ast.Analysis
+  alias Lexical.Document
+  alias Lexical.Document.Edit
+  alias Lexical.Document.Line
+  alias Lexical.Document.Position
+  alias Lexical.RemoteControl.CodeIntelligence.Entity
+  alias Lexical.RemoteControl.Search.Store
+  require Logger
+
+  import Line
+
+  @spec rename(Analysis.t(), Position.t(), String.t()) ::
+          {:ok, %{Lexical.uri() => [Edit.t()]}} | {:error, term()}
+  def rename(%Analysis{} = analysis, %Position{} = position, new_name) do
+    with {:ok, entity, range} <- resolve_module(analysis, position) do
+      edits =
+        analysis.document
+        |> search_related_candidates(position, entity, range)
+        |> to_uri_with_changes(new_name)
+
+      {:ok, edits}
+    end
+  end
+
+  defp resolve_module(analysis, position) do
+    case Entity.resolve(analysis, position) do
+      {:ok, {module_or_struct, module}, range} when module_or_struct in [:struct, :module] ->
+        {:ok, module, range}
+
+      {:ok, other, _} ->
+        {:error, {:unsupported_entity, other}}
+
+      {:error, reason} ->
+        {:error, reason}
+    end
+  end
+
+  defp search_related_candidates(document, position, entity, range) do
+    cursor_entity_string = cursor_entity_string(range)
+
+    entities =
+      exacts(entity, cursor_entity_string)
+
+    # Users won't always want to rename descendants of a module.
+    # For instance, when there are no more submodules after the cursor.
+    # like: `defmodule TopLevel.Mo|dule do`
+    # in most cases, users only want to rename the module itself.
+    #
+    # However, there's an exception when the cursor is in the middle,
+    # such as `Top.Mo|dule.ChildModule`. If we rename it to `Top.Renamed.Child`,
+    # it would be natural to also rename `Module.ChildModule` to `Renamed.Child`.
+    if at_the_middle_of_module?(document, position, range) do
+      entities ++ descendants(entity, cursor_entity_string)
+    else
+      entities
+    end
+  end
+
+  defp at_the_middle_of_module?(document, position, range) do
+    range_text = range_text(range)
+
+    case Ast.surround_context(document, position) do
+      {:ok, %{context: {:alias, alias}}} ->
+        String.length(range_text) < length(alias)
+
+      _ ->
+        false
+    end
+  end
+
+  defp cursor_entity_string(range) do
+    # Parent.|Module -> Module
+    range
+    |> range_text()
+    |> String.split(".")
+    |> List.last()
+  end
+
+  defp descendants(entity, cursor_entity_string) do
+    entity_string = inspect(entity)
+    prefix = "#{entity_string}."
+
+    case Store.prefix(prefix, subject: [:definition, :reference]) do
+      {:ok, results} ->
+        filtered =
+          results
+          |> Enum.filter(fn result ->
+            range_text = range_text(result.range)
+            String.contains?(range_text, cursor_entity_string)
+          end)
```

Comment:
This particular filter happens at least three times in this file, I think -- probably worth extracting into a helper.

```suggestion
          |> entries_matching(cursor_entity_string)
```

Reply:
haha, Just twice, but I prefer to abstract a function for the item